### DISCUSSION TITLE
Can't Get ruTorrent To Connect to rTorrent Using Basic Instructions

### DISCUSSION THREAD

**Discussion by drak01** [Novik/ruTorrent#2730] [Q&A]
I'm pulling my hair out trying to get my ruTorrent to connect to rTorrent.  I just upgraded to Ubuntu 24.10 and I'm trying to transfer everything over.  I'm starting with the absolute basics with basic configuration.

1) I installed rtorrent via `sudo apt-get install rtorrent`  Version is 0.9.8.  I downloaded .rtorrent.rc from `https://github.com/rakshasa/rtorrent/wiki/CONFIG-Template` by copy and pasting the default config.  rTorrent works and is running.  However, when running `ss -ltnp` I only see 

    `LISTEN    0         4096                     *:55556                  *:*        users:(("rtorrent main",pid=21693,fd=14))` 

    I don't see any port 5000 SCGI listening port.

2) I went to `https://github.com/Novik/ruTorrent/wiki/WebSERVER`. I installed apache2 via `sudo apt-get install apache2` .  I installed `libapache2-mod-php`, `php-mbstring`, `php-cgi`, `php-fpm`.  I can't remember what else I installed, the instructions aren't super clear.  

3) When running `sudo apt-get install libapache2-mod-scgi` I'm told `E: Package 'libapache2-mod-scgi' has no installation candidate`. It looks like scgi is deprecated.  I followed the instructions [here](https://github.com/Bercik1337/rt-auto-install/issues/7) and downloaded the deb of 1.13 for amd64.  Hopefully this is ok.

4) I ran `sudo a2enmod scgi` and the same for php8.3.  It says they are enabled.  I edited `/etc/apache2/sites-enabled/000-default.conf` and add `SCGIMount /RPC2 127.0.0.1:5000` inside the `<VirtualHost *:80>` directive.  I ran `sudo a2ensite 000-default.conf`.  It's enabled.  I restart my webserver constantly with `sudo systemctl restart apache2.service`.  I check phpinfo() and can see mbstring and mod_scgi.

5) I download the stable ruTorrent-5.0.0 tar.gz.  I move it over to `/var/www/html` and `sudo tar -zxvf ruTorrent-5.0.0.tar.gz`.  I change ownership of the files from root to www-data using `sudo chown -R www-data:www-data ruTorrent-5.0.0` and rename to just ruTorrent.  In `config.php` the scgi_port and scgi_host is 5000 and 127.0.0.1 so I don't touch anything.  I go to 127.0.0.1/ruTorrent.  The interface pops up, but says `Torrent list not available yet, connection to rTorrent not established`.  I check `/var/log/apache2/error.log` and everything there looks fine.  I try to check `/tmp/errors.log` for the ruTorrent log but that file doesn't exist.

6) I try Mod Proxy SCGI.  I edited `/etc/apache2/sites-enabled/000-default.conf` and replaced the old `SCGIMount /RPC2 127.0.0.1:5000` with `ProxyPass /RPC2 scgi://localhost:5000/` and `sudo a2enmod proxy_scgi` and restarted the server.  I refresh ruTorrent and it says the same thing.

7) I try the unix socket.  I replace `ProxyPass /RPC2 scgi://localhost:5000/` with `ProxyPass /RPC2 unix:/tmp/rpc.socket|scgi://127.0.0.1` and edited config.php by commenting out scgi_host and scgi_port and uncommenting the socket ($scgi_port = 0; $scgi_host = "unix:///tmp/rpc.socket";)  Still doesn't work.

Without going into all the details I've tried lighttpd and nginx as well even though I'm less familiar with those.  I haven't tried the HTTPRPC plugin, I'm not sure how to install that.  Doesn't anybody have any ideas?  I'm surely missing something simple.

I think it might be something with scgi.  When I tried `xmlrpc http://127.0.0.1/RPC2 view.list` I get an error message and my apache log says

`[Sat Oct 12 18:24:08.973154 2024] [:error] [pid 44089:tid 44089] (111)Connection refused: [client 127.0.0.1:58788] scgi: scgi: can't connect to server`

error message:

`Failed.  Call failed.  HTTP POST to URL 'http://127.0.0.1/RPC2' failed.  HTTP response code is 503, not 200.  (XML-RPC fault code -504)`

**Comment by drak01** [comment]
Ok, I installed lighttpd again with RPC host and port and I think I made progress.  Now it's giving me this error when I launch ruTorrent:

`[12.10.2024 20:56:57] WebUI started.[12.10.2024 20:56:57] Bad response from server: (403 [error,getuisettings]) <!DOCTYPE html> <html lang="en"> <head> <meta charset="UTF-8" /> <title>403 Forbidden</title> </head> <body> <h1>403 Forbidden</h1> </body> </html>`
`[12.10.2024 20:56:57] Bad response from server: (403 [error,getplugins]) Forbidden`

**Comment by drak01** [comment]
I'm getting some weird 403 Forbidden when it seems I'm accessing my php files, but when I access `index.lighttpd.html` in the web root directory I can read that file just fine.

**Comment by stickz** [comment]
Hello, I would recommend the swizzin installation script on **Ubuntu 24.04 LTS**. This type of installation method is complex and time consuming. You can be up and running in 10 minutes on rTorrent 0.10.0 with ruTorrent v5.0. https://github.com/swizzin/swizzin

**Comment by drak01** [comment]
Ty stickz, I can use the script as a last resort, but I'd like to figure out how this works.

I'm trying now the nginx route.  I installed nginx and php-fpm using [this guide](https://www.digitalocean.com/community/tutorials/php-fpm-nginx).  I can load ruTorrent and it seems to connect to rTorrent, but it's giving me the following

`Bad response from server: (500 [error,getplugins]) Internal Server Error`

Checking nginx error logs I see:

`2024/10/13 15:52:41 [error] 212761#212761: *84 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Uncaught Error: Call to undefined function exec() in /var/www/html/ruTorrent/php/which.php:30`
`Stack trace:`
`#0 /var/www/html/ruTorrent/php/which.php(15): WhichCache->setFilePath()`
`#1 /var/www/html/ruTorrent/php/which.php(86): WhichCache->getFilePath()`
`#2 /var/www/html/ruTorrent/php/getplugins.php(383): findEXE()`
`#3 {main}`
`  thrown in /var/www/html/ruTorrent/php/which.php on line 30" while reading response header from upstream, client: 127.0.0.1, server: _, request: "GET /ruTorrent/php/getplugins.php HTTP/1.1", upstream: "fastcgi://unix:/var/run/php/php-fpm.sock:", host: "127.0.0.1", referrer: "http://127.0.0.1/ruTorrent/"`

When I check phpinfo() I see under Core

`disable_functions	exec,passthru,shell_exec,system	exec,passthru,shell_exec,system`

Yet when I check `Loaded Configuration File 	/etc/php/8.3/fpm/php.ini `

`; This directive allows you to disable certain functions.`
`; It receives a comma-delimited list of function names.`
`; https://php.net/disable-functions`
`disable_functions =`

disable_functions is blank... but for some reason my exec function is disabled..

**Comment by drak01** [comment]
Some more info..

when I run phpinfo() through the command line php command `disable_function` is blank and when I test another script that checks for exec it says it's enabled..

**Comment by drak01** [comment]
Ah, I had 

`php_admin_value[disable_functions] = exec,passthru,shell_exec,system` 

set in one of my conf files in the `/etc/php/8.3/fpm/pool.d/` directory.  Will have to look further.

Error message:

`2024/10/13 21:54:21 [crit] 31604#31604: *1 connect() to unix:/var/run/php8.3-fpm.sock failed (2: No such file or directory) while connecting to upstream, client: 127.0.0.1, server: _, request: "GET /ruTorrent/php/getplugins.php HTTP/1.1", upstream: "fastcgi://unix:/var/run/php8.3-fpm.sock:", host: "127.0.0.1", referrer: "http://127.0.0.1/ruTorrent/"`

Looks like some permissions issue.

**Comment by jevenski** [comment]
It's definitely a file permission issue. The user running the web server (usually `www-data`) will need to have read access to the unix domain socket, which is created and owned by the user running rtorrent. If you add `www-data` to rtorrent user's user group in order for it to be able to read the socket, you'll need to reboot the system to take the change into effect.

**Comment by drak01** [comment]
Does rTorrent create the socket or php-fpm?  It seems like php-fpm creates the socket and the one I'm running is in `/var/run/php/php8.3-fpm.sock`.  

`t1@TOWERLIN:/var/run/php$ ls -lsa`
`total 4`
`0 drwxr-xr-x  2 www-data www-data  100 Oct 14 14:26 .`
`0 drwxr-xr-x 40 root     root     1000 Oct 14 14:26 ..`
`4 -rw-r--r--  1 root     root        4 Oct 14 14:26 php8.3-fpm.pid`
`0 srw-rw----  1 www-data www-data    0 Oct 14 14:26 php8.3-fpm.sock`
`0 lrwxrwxrwx  1 root     root       30 Oct 14 14:26 php-fpm.sock -> /etc/alternatives/php-fpm.sock`

The socket is owned by www-data as is the process `php-fpm` and `nginx`.

`t1@TOWERLIN:/var/run/php$ ps aux | grep php`
`root        1754  0.0  0.1 233204 28120 ?        Ss   14:26   0:00 php-fpm: master process (/etc/php/8.3/fpm/php-fpm.conf)`
`www-data    1784  0.0  0.1 233800 20924 ?        S    14:26   0:00 php-fpm: pool www`
`www-data    1785  0.0  0.1 233800 20816 ?        S    14:26   0:00 php-fpm: pool www`
`t1          6649  0.0  0.0  17888  2476 pts/1    S+   14:37   0:00 grep --color=auto php`

`t1@TOWERLIN:/var/run/php$ ps aux | grep nginx`
`root       10602  0.0  0.0  13760  1872 ?        Ss   15:04   0:00 nginx: master process /usr/sbin/nginx -g daemon on; master_process on;`
`www-data   10603  0.0  0.0  15864  6028 ?        S    15:04   0:00 nginx: worker process`
`www-data   10604  0.0  0.0  15568  4864 ?        S    15:04   0:00 nginx: worker process`
`www-data   10605  0.0  0.0  15568  5616 ?        S    15:04   0:00 nginx: worker process`
`www-data   10606  0.0  0.0  15568  4828 ?        S    15:04   0:00 nginx: worker process`
`t1         10717  0.0  0.0  17888  2476 pts/1    S+   15:05   0:00 grep --color=auto nginx`

The user running rtorrent is t1.  He is already a member of www-data.

`t1@TOWERLIN:/var/run/php$ sudo adduser t1 www-data`
```info: The user `t1' is already a member of `www-data'.```

In `/etc/php/8.3/fpm/pool.d` there is only one conf now, `www.conf`.  Some excerpts include,

`[www]`
`user = www-data`
`group = www-data`
`listen = /run/php/php8.3-fpm.sock`
`listen.owner = www-data`
`listen.group = www-data`

There's a lot of moving pieces.  I'm sorry I don't have a fuller understanding.  How does rtorrent know where to use the socket?

Also, my logs rotated.  I'm not getting any errors anymore in `/var/log/nginx/error.log`  when I go to `http://127.0.0.1/ruTorrent` but it's saying `"Torrent list not yet available, connection to rTorrent not established."`

Yeah I'm just not understanding all the connections.  I think I just have to get rtorrent to connect to the socket now.

**Comment by drak01** [comment]
So it looks like rtorrent opens a socket here (from `.rtorrent.rc`)

`network.scgi.open_local = (cat,(session.path),rpc.socket)`

which is `/home/t1/rtorrent/.session/rpc.socket`

When I change this to `network.scgi.open_local = /run/php/php8.3-fpm.sock` it says `rtorrent: Error in option file: ~/.rtorrent.rc:111: Could not prepare socket for listening: Address already in use`

I guess I don't understand how the sockets work.

**Comment by n1ce-0ne** [comment]
just a guy reading along taking an interest & willing to help if i can.
i know you installed rtorrent & say it was running but you didnt mention how you set it running & mention no listening port.
did you run rtorrent in screen yea? screen -dmS rtorrent rtorrent
is it still running ? screen -ls

worth checking

just seen your last comment..

weird i dont have network.scgi.open_local in my rc
only network.scgi.open related option i have is network.scgi.open_port = 127.0.0.1:5002

**Comment by drak01** [comment]
Ah, I had it uncommented when it's normally commented.  I'm not sure what configuration I need so rtorrent is setup to talk to the webserver/rutorrent.  Docs seems to say to go [here](https://github.com/rakshasa/rtorrent/wiki/CONFIG-Template) so I downloaded the basic .rtorrent.rc from there and just trying to work from a default installation.  I'm trying to start from basics.  Just a simple rtorrent, then install web server, then drop in rutorrent, then install what I need to get the webserver to talk to rtorrent.

I'm not running screen or anything advanced yet.  I'm just running rtorrent from my terminal.

There seems to be a lot of different options to run rutorrent through the webserver and get it to talk to rtorrent.  I'm just trying to find one that works lol.  I thought I would know enough, but evidently I need to know more to understand and get the job done.  I used to just use apache and scgi on ubuntu 15.10, but that doesn't seem to work anymore, or at least I'm not up to date on the new changes.

I tried going the scgi route with rpc host and port, but I had problems.  I can try again, but without jumping all over the place I'd like to go step by step and understand each step and work things through.

**Comment by n1ce-0ne** [comment]
i feel ya m8, its pretty darn interesting stuff init hahaa one reason i enjoy reading & learning here.
imho i'd say give the scgi route with rpc host and port another shot, this method does seem the default method in most scripts ive used

**Comment by drak01** [comment]
I seemed to have fixed the problem by making sure I have 

`network.scgi.open_port = 127.0.0.1:5000` in my `.rtorrent.rc` file.  (I commented out `#network.scgi.open_local` just in case)

Adding 

`location /RPC2 {`
                `include scgi_params;`
                `scgi_pass 127.0.0.1:5000;`
        `}`

In my `/etc/nginx/sites-enabled/default` .

Now I'm getting

`[19.10.2024 16:10:11] _cloudflare: Plugin will not work. Webserver user can't access external program (python).`
`[19.10.2024 16:10:11] rutracker_check: Plugin will not work. Webserver user must have execute access to the rtorrent session directory (/home/t1/rtorrent/.session/).`
`[19.10.2024 16:10:11] dump: Plugin will not work. rTorrent user can't access external program (dumptorrent).`
`[19.10.2024 16:10:11] mediainfo: Plugin will not work. rTorrent user can't access external program (mediainfo).`
`[19.10.2024 16:10:11] spectrogram: Plugin will not work. rTorrent user can't access external program (sox).`
`[19.10.2024 16:10:12] Unpack plugin: rTorrent user can't access 'unrar' program.`

I had the same error for ffmpeg, but I solved it by simply installing it so that seems to be the fix.  So that's what I'm going to do and it should fix my problems.  I think I might ignore the rutracker_check webserver permission issue since I don't use it.  Hopefully it won't effect my usage.  

If I have time I may write a tutorial.  We shall see if everything works.

**Comment by drak01** [comment]
Everything's working so I will close this.  Thanks everyone.  :)
